File: Algo/Leetcode/208TrieTree.cpp
struct TrieNode{
    bool isEnd;
    struct TrieNode* ptr[26];
    TrieNode(bool e): isEnd(e) {
        memset(ptr, 0, sizeof(ptr));
    }
};
class Trie {
public:
    /** Initialize your data structure here. */
    Trie() {
        root = new TrieNode(true);
    }

    /** Inserts a word into the trie. */
    void insert(string word) {
        insert(root, word);
    }

    void insert(TrieNode* root, string word){
        int n = word.length();
        if(n==0) return;
        int idx = word[0] - 'a';
        if(!root->ptr[idx]) root->ptr[idx] = new TrieNode(false);
        if(n==1){
            root->ptr[idx]->isEnd = true;
            return;
        }
        insert(root->ptr[idx], word.substr(1));
    }

    /** Returns if the word is in the trie. */
    bool search(string word) {
        return search(root, word);
    }

    bool search(TrieNode* root, string word){
        int n = word.length();
        if(n==0) return root->isEnd;
        if(!root) return false;
        int idx = word[0] - 'a';
        if(!root->ptr[idx]) return false;
        return search(root->ptr[idx], word.substr(1));
    }

    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
        return startsWith(root, prefix);
    }

    bool startsWith(TrieNode* root, string prefix) {
        int n = prefix.length();
        if(n==0) return true;
        if(!root) return false;
        int idx = prefix[0] - 'a';
        if(!root->ptr[idx]) return false;
        return startsWith(root->ptr[idx], prefix.substr(1));
    }

    TrieNode* root;
};

File: Algo/Leetcode/208TrieTree.py
class Node(object):

    def __init__(self, v=''):
        self.v = v
        self.ptr = [None] * 26
        self.left = None
        self.right = None
        self.endNode = False


class Trie(object):

    def __init__(self):
        self.root = Node()

    def insert(self, word):
        """
        :type word: str
        :rtype: None
        """
        return self.insertNode(self.root, str(word))

    def insertNode(self, node, word):
        if len(word) == 0 and node is not None:
            node.endNode = True
            return None

        c = word[0]
        idx = self.ord(c)
        if node.ptr[idx] is None:
            node.ptr[idx] = Node(c)

        self.insertNode(node.ptr[idx], word[1:])
        return None

    def search(self, word):
        """
        :type word: str
        :rtype: bool
        """
        return self.searchNode(self.root, str(word))

    def searchNode(self, node, word):
        if len(word) == 0:
            if node.endNode:
                return True
            else:
                return False

        if node is None:
            return False

        c = word[0]
        idx = self.ord(c)
        if node.ptr[idx] is None:
            return False

        return self.searchNode(node.ptr[idx], word[1:])

    def startsWith(self, prefix):
        """
        :type prefix: str
        :rtype: bool
        """
        return self.startsWithNode(self.root, str(prefix))

    def startsWithNode(self, node, prefix):
        if len(prefix) == 0:  # no need to be leaf node
            return True

        c = prefix[0]
        idx = self.ord(c)
        if node.ptr[idx] is None:
            return False

        return self.startsWithNode(node.ptr[idx], prefix[1:])

    def ord(self, c):
        return ord(c) - ord('a')

# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)